borsh/src/schema/container_ext/max_size.rs

```diff
@@ -0,0 +1,416 @@
+        Ok(Definition::Array { length, elements }) => {
+            *length == 0 || is_zero_size_impl(elements.as_str(), schema, stack)?
+        }
+        Ok(Definition::Sequence { .. }) => false,
+        Ok(Definition::Tuple { elements }) => all(elements.iter(), |key| *key, schema, stack)?,
+        Ok(Definition::Enum {
+            tag_width: 0,
+            variants,
+        }) => all(
+            variants.iter(),
+            |(_variant_name, declaration)| declaration,
+            schema,
+            stack,
+        )?,
+        Ok(Definition::Enum { .. }) => false,
+        Ok(Definition::Struct { fields }) => match fields {
+            Fields::NamedFields(fields) => all(
+                fields.iter(),
+                |(_field_name, declaration)| declaration,
+                schema,
+                stack,
+            )?,
+            Fields::UnnamedFields(fields) => {
+                all(fields.iter(), |declaration| declaration, schema, stack)?
+            }
+            Fields::Empty => true,
+        },
+
+        Err(dec) => dec == "nil",
+    };
+    stack.pop();
+    Ok(res)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // this is not integration test module, so can use __private for ease of imports;
+    // it cannot be made integration, as it tests `is_zero_size` function, chosen to be non-pub
+    use crate::{
+        BorshSchema,
+        __private::maybestd::{
+            boxed::Box,
+            collections::BTreeMap,
+            format,
+            string::{String, ToString},
+            vec,
+        },
+    };
+
+    #[track_caller]
+    fn test_ok<T: BorshSchema>(want: usize) {
+        let schema = BorshSchemaContainer::for_type::<T>();
+        assert_eq!(Ok(want), schema.max_serialized_size());
+        assert_eq!(
+            want == 0,
+            is_zero_size(schema.declaration(), &schema).unwrap()
+        );
+    }
+
+    #[track_caller]
+    fn test_err<T: BorshSchema>(err: MaxSizeError) {
+        let schema = BorshSchemaContainer::for_type::<T>();
+        assert_eq!(Err(err), schema.max_serialized_size());
+    }
+
+    const MAX_LEN: usize = u32::MAX as usize;
+
+    #[test]
+    fn test_is_zero_size_recursive_check_bypassed() {
+        use crate as borsh;
+
+        #[derive(::borsh_derive::BorshSchema)]
+        struct RecursiveExitSequence(Vec<RecursiveExitSequence>);
+
+        let schema = BorshSchemaContainer::for_type::<RecursiveExitSequence>();
+        assert_eq!(Ok(false), is_zero_size(schema.declaration(), &schema));
+    }
+
+    #[test]
+    fn test_is_zero_size_recursive_check_err() {
+        use crate as borsh;
+
+        #[derive(::borsh_derive::BorshSchema)]
+        struct RecursiveNoExitStructUnnamed(Box<RecursiveNoExitStructUnnamed>);
+
+        let schema = BorshSchemaContainer::for_type::<RecursiveNoExitStructUnnamed>();
+        assert_eq!(Err(()), is_zero_size(schema.declaration(), &schema));
+    }
```

Comment:
a similar test in current master was causing stack overflow
```rust
    #[test]
    fn test_is_zero_size_recursive_check_err() {
        use crate as borsh;

        #[derive(::borsh_derive::BorshSchema)]
        struct RecursiveNoExitStructUnnamed(Box<RecursiveNoExitStructUnnamed>);

        let schema = BorshSchemaContainer::for_type::<RecursiveNoExitStructUnnamed>();
        assert_eq!(false, is_zero_size(schema.declaration(), &schema));
    }
```
```
thread 'schema_helpers::tests::test_is_zero_size_recursive_check_err' has overflowed its stack
fatal runtime error: stack overflow
```
Though rustc prevents from constructing a type, which would hit a [branch with recursing indefinetely into `is_zero_size`](https://github.com/near/borsh-rs/blob/master/borsh/src/schema_helpers.rs#L189):

```bash
error: values of the type `[[[RecursiveNoExitStructUnnamed2; 4294967295]; 2]; 4294967295]` are too big for the current architecture
```